### ISSUE TITLE
Request for more details

### ISSUE THREAD

**Issue by KennyNH** [black-yt/Exformer#1]
Thanks for your wonderful work. 

Could you further release the implementation of extreme value metrics (i.e., RQE and SEDI), and the detailed codes for model training? I believe it will be much easier to follow your work.

Thank you for taking time out of your busy schedule to read my email. Please forgive me for any inconvenience! Hope reply ~

**Comment by black-yt** [comment]
Thank you for your suggestion, we have added test code for RQE and SEDI. Regarding the complete model code, we will make it publicly available after the article is accepted.

**Comment by KennyNH** [comment]
Thanks for your effort and valuable time. Besides, could you please provide the source code of FuXi-Extreme? Since I fail to find the code.

**Comment by black-yt** [comment]
As far as we know, FuXi-Extreme currently does not have open source code, and the SEDI we show in the paper is derived from its paper https://arxiv.org/pdf/2310.19822.pdf.

**Comment by KennyNH** [comment]
Thanks for your response. I have two further questions about the implementation of RQE and SEDI, when we have a large testing dataset and are required to calculate the metrics for each batch, formulated as shape=[B (batch size), T (num of lead time steps), H, W, C (num of variables)].

Regarding SEDI: 
(1) How to define the quantile threshold? For each point in the THWC-shaped data, is the quantile operation conducted within each batch to get batch-specific quantile thresholds, and we average the SEDI values of all the batches for the overall SEDI? Or do we have to beforehand conduct the quantile operation within the entire testing data (i.e., covering all the batches), and use the same quantile thresholds for each batch to compute the SEDI value?
(2) Are the quantile thresholds defined with the ground-truth target values or the predicted target values?

Regarding RQE: 
A similar question. Do we need to compute the RQE within each THWC-shaped sample? Or within each batch? Or within the entire testing data?

**Comment by KennyNH** [comment]
Besides, I have a question about the curves in Figure 4,5,6 of your paper. Regarding the performance of each lead time point in the horizontal axis, do we need to average the results of all the lead time steps? Or we just use the result of the specific lead time step. E.g., if we want to plot the RQE point for variable u10 and lead time 25, we first compute the RQE result which is a (25,)-shaped tesnor **_R_**. Then to plot the point, should we use the **_R[24]_** or the **_Mean(R)_** as the Y-axis value?

**Comment by black-yt** [comment]
The values obtained by SEDI and RQE during batch evaluation are the average of B samples in the batch. If you want to calculate SEDI and RQE for all samples in the entire dataset, you need to average all batches again.

The threshold of SEDI comes from the real data of ERA5. Specifically, the corresponding quantiles are obtained by counting the data from 1993 to 2016. For specific methods, please refer to "3.3.2 Symmetric extremal dependency index (SEDI)" in FuXi-Extreme.

For different lead times, such as 6h and 24h, SEDI and RQE should be calculated separately, and there is no need to average between different lead times.

**Comment by KennyNH** [comment]
Thanks for your response. It is really helpful.

(1) I feel confused because I find that in the code of Earthformer, the CSI is calculated once for all the test samples among all the test batches, rather than calculated for each sample and averaged among all the test batches (as you mentioned). I think the implementation of SEDI should be similar. So I wonder whether the distinction of implementation affects the accurate evaluation of extreme forecasting or not. Would you mind sharing any insights on that?

(2) I just want to have a check. If I'm not mistaken, the thresholds are computed with the raw weather data, rather than the spatio-temporal sequences processed by sliding window. Is that correct?

**Comment by black-yt** [comment]
In fact, if your device has enough memory, you can indeed calculate the indicators for all samples in the test set at once and average them. It is the same as averaging the batch. For example (1+3+3+2+1+2)/6=((1+3+3)/3+(2+1+2)/3)/2.

Perhaps calculating CSI in some smaller datasets, such as precipitation forecasts, involves the time dimension. For medium-range forecasts, SEDI and RQE will not have a T dimension in their calculations, but a single time. And no matter what lead time is, the SEDI threshold is fixed. 

The figure of SEDI in our paper is the summary results after calculating the SEDI of forecasting at different lead times respectively.

**Comment by KennyNH** [comment]
Thanks for your reply.

You may misunderstood my first question. For example, If we want to compute the CSI for a test dataset with two batches. Your implementation could be formulated as `CSI = (CSI_1 + CSI_2) / 2 = (TP_1 / (TP_1 + FN_1 + FT_1) + TP_2 / (TP_2 + FN_2 + FT_2)) / 2`. But the implementation of Earthformer is `CSI = (TP_1 + TP_2) / ((TP_1 + TP_2) + (FN_1 + FN_2) + (FT_1 + FT_2))`. Thus the results of these two implementations are distinct. This difference also exists in the implementation of SEDI and RQE. My question is which one is preferable.

Note: `TP_1, TP_2` denote the `#True Positives` respectively in batch 1 and batch 2.

**Comment by black-yt** [comment]
Thank you for pointing out the problem. We have modified the code so that SEDI calculates each sample in the batch separately and then averages it. The new code has been committed to the repository.

Note: Since the batch size we used when testing in the paper is 1, the above problems will not affect the indicators in the paper.

**Comment by KennyNH** [comment]
Regarding Line 66-72, 
```
F = FP/(FP+TN)
H = TP/(TP+FN)

SEDI = (torch.log(F)-torch.log(H)-torch.log(1-F)+torch.log(1-H))/ \
       (torch.log(F)+torch.log(H)+torch.log(1-F)+torch.log(1-H))

SEDI = torch.mean(SEDI, dim=0)
```
Could we change it as follows:
```
FP, TN, TP, FN = FP.sum(), TN.sum(), TP.sum(), FN.sum()

F = FP/(FP+TN)
H = TP/(TP+FN)

SEDI = (torch.log(F)-torch.log(H)-torch.log(1-F)+torch.log(1-H))/ \
       (torch.log(F)+torch.log(H)+torch.log(1-F)+torch.log(1-H))
```
where the SEDI seems more comprehensive, as `FP, TN, TP, FN` are counted with all the samples rather than a single sample separately.

**Comment by black-yt** [comment]
I think the first method is better, that is, calculate each sample first, and then average the samples in the batch (SEDI = torch.mean(SEDI, dim=0)).

**Comment by KennyNH** [comment]
OK, I get it. Thanks for your timely response.
